pex/resolver.py

```diff
@@ -95,6 +126,401 @@ def parsed_platform(platform=None):
+    # 2. Build local projects and sdists.
+    # 3. Install wheels in individual chroots.
+    # 4. Calculate the final resolved requirements.
+    #
+    # You'd think we might be able to just pip install all the requirements, but pexes can be
+    # multi-platform / multi-interpreter, in which case only a subset of distributions resolved into
+    # the PEX should be activated for the runtime interpreter. Sometimes there are platform specific
+    # wheels and sometimes python version specific dists (backports being the common case). As such,
+    # we need to be able to add each resolved distribution to the `sys.path` individually
+    # (`PEXEnvironment` handles this selective activation at runtime). Since pip install only
+    # accepts a single location to install all resolved dists, that won't work.
+    #
+    # This means we need to seperately resolve all distributions, then install each in their own
+    # chroot. To do this we use `pip download` for the resolve and download of all needed
+    # distributions and then `pip install` to install each distribution in its own chroot.
+    #
+    # As a complicating factor, the runtime activation scheme relies on PEP 425 tags; i.e.: wheel
+    # names. Some requirements are only available or applicable in source form - either via sdist,
+    # VCS URL or local projects. As such we need to insert a `pip wheel` step to generate wheels for
+    # all requirements resolved in source form via `pip download` / inspection of requirements to
+    # discover those that are local directories (local setup.py or pyproject.toml python projects).
+    #
+    # Finally, we must calculate the pinned requirement corresponding to each distribution we
+    # resolved along with any environment markers that control which runtime environments the
+    # requirement should be activated in.
+
+    if not self._requirements and not self._requirement_files:
+      # Nothing to resolve.
+      return []
+
+    workspace = safe_mkdtemp()
+    cache = self._cache or workspace
+
+    resolved_dists_dir = os.path.join(workspace, 'resolved_dists')
+    spawn_resolve = functools.partial(self._spawn_resolve, resolved_dists_dir)
+    to_resolve = self._targets
+
+    built_wheels_dir = os.path.join(cache, 'built_wheels')
+    spawn_wheel_build = functools.partial(self._spawn_wheel_build, built_wheels_dir)
+    to_build = list(self._iter_local_projects())
+
+    installed_wheels_dir = os.path.join(cache, 'installed_wheels')
+    spawn_install = functools.partial(self._spawn_install, installed_wheels_dir)
+    to_install = []
+
+    to_calculate_requirements_for = []
+
+    # 1. Resolve sdists and wheels.
+    with TRACER.timed('Resolving for:\n  '.format('\n  '.join(map(str, to_resolve)))):
+      for resolve_result in self._run_parallel(inputs=to_resolve,
+                                               spawn_func=spawn_resolve,
+                                               raise_type=Unsatisfiable):
```

Comment:
Agreed on all points. Note though the shape of the steps is roughly:
+ downloads: 1-4 
+ builds: 1-10
+ installs: 1-100

For downloads, you may have 1-2 interpreters (python 2.7 + python 3) and -2 platforms.
For installs projects pulling in a transitive dep graph of 100 wheels is not unreasonable, but things probably don't get alot bigger than this either.
For builds, this will be a subset of the installs and it seems like (handwave), most pypi projects have prebuilt wheels for standard platforms and a 10% of my dependencies require local builds is about right for the repos I have access to.

Reply:
Thank you for this context!!